File: exercicios/projeto-final/skeletons/pokeapi/kmp/pratica/androidApp/src/main/kotlin/com/puciec/pokedexkmp/ListScreen.kt
package com.puciec.pokedexkmp

import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.LazyRow
import androidx.compose.foundation.lazy.items
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.FilterChip
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.OutlinedTextField
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.testTag
import androidx.compose.ui.unit.dp
import com.puciec.pokedexkmp.data.PokeApi
import com.puciec.pokedexkmp.data.PokemonSummary
import com.puciec.pokedexkmp.data.capitalize

private val CATEGORIES = listOf("fire", "water", "electric", "grass")

@Composable
fun ListScreen(api: PokeApi, onSelect: (Int) -> Unit) {
    var all by remember { mutableStateOf<List<PokemonSummary>>(emptyList()) }
    var loading by remember { mutableStateOf(true) }
    var error by remember { mutableStateOf<String?>(null) }
    var searchText by remember { mutableStateOf("") }
    var selectedCategory by remember { mutableStateOf<String?>(null) }
    var categoryNames by remember { mutableStateOf<Set<String>?>(null) }

    LaunchedEffect(Unit) {
        try {
            all = api.fetchList()
        } catch (e: Exception) {
            error = e.message ?: "Erro desconhecido"
        } finally {
            loading = false
        }
    }

    val filtered = all.filter {
        (searchText.isBlank() || it.name.contains(searchText, ignoreCase = true))
                && (selectedCategory == null || categoryNames?.contains(it.name) == true)
    }

    if (loading) {
        Box(Modifier.fillMaxSize(), contentAlignment = Alignment.Center) { CircularProgressIndicator() }
        return
    }
    if (error != null) {
        Box(Modifier.fillMaxSize(), contentAlignment = Alignment.Center) { Text(error!!) }
        return
    }

    Box(Modifier.fillMaxSize().testTag("item-list-screen")) {
        LazyColumn(Modifier.fillMaxSize()) {
            item {
                OutlinedTextField(
                    value = searchText,
                    onValueChange = { searchText = it },
                    label = { Text("Buscar pokémon") },
                    modifier = Modifier.fillMaxWidth().padding(12.dp).testTag("search-input"),
                )
            }
            item {
                LazyRow(
                    Modifier.fillMaxWidth().padding(horizontal = 12.dp),
                    horizontalArrangement = Arrangement.spacedBy(8.dp),
                ) {
                    item {
                        FilterChip(
                            selected = selectedCategory == null,
                            onClick = { selectedCategory = null; categoryNames = null },
                            label = { Text("Todos") },
                            modifier = Modifier.testTag("category-chip-all"),
                        )
                    }
                    items(CATEGORIES) { category ->
                        FilterChip(
                            selected = selectedCategory == category,
                            onClick = { selectedCategory = category },
                            label = { Text(capitalize(category)) },
                            modifier = Modifier.testTag("category-chip-$category"),
                        )
                    }
                }
            }
            items(filtered, key = { it.id }) { pokemon ->
                Text(
                    text = capitalize(pokemon.name),
                    modifier = Modifier
                        .fillMaxWidth()
                        .testTag("item-card-${pokemon.id}")
                        .clickable { onSelect(pokemon.id) }
                        .padding(16.dp),
                    style = MaterialTheme.typography.bodyLarge,
                )
            }
        }
    }

    LaunchedEffect(selectedCategory) {
        val category = selectedCategory ?: return@LaunchedEffect

        try {
            categoryNames = api.fetchNamesByType(category)
        } catch (e: Exception) {
            error = e.message ?: "Erro desconhecido"
        }
    }
}

File: exercicios/projeto-final/skeletons/pokeapi/flutter/pratica/lib/screens/list_screen.dart
import 'package:flutter/material.dart';
import '../api/poke_api.dart';
import '../models/pokemon.dart';
import 'detail_screen.dart';

const List<String> _categories = ['fire', 'water', 'electric', 'grass'];

class ListScreen extends StatefulWidget {
  const ListScreen({super.key});

  @override
  State<ListScreen> createState() => _ListScreenState();
}

class _ListScreenState extends State<ListScreen> {
  final PokeApi _api = PokeApi();

  List<PokemonSummary> _all = [];
  bool _loading = true;
  String? _error;
  String _searchText = '';
  String? _selectedCategory;
  Set<String>? _categoryNames;

  @override
  void initState() {
    super.initState();
    _loadList();
  }

  Future<void> _loadList() async {
    // TODO 1 (feature 1 — lista): chamar `_api.fetchList()`, guardar o resultado
    // em `_all` e marcar `_loading = false` via setState. Tratar erro com
    // `_error` (mensagem) igual ao padrão dos outros catch deste arquivo.
    setState(() {
      _loading = false;
    });
  }

  Future<void> _selectCategory(String? category) async {
    // TODO 4 (feature 4 — categoria/filtro): quando `category` não é null,
    // chamar `_api.fetchNamesByType(category)` e guardar em `_categoryNames`
    // (setState). Quando `category` é null, limpar `_categoryNames`.
    setState(() {
      _selectedCategory = category;
    });
  }

  List<PokemonSummary> get _filtered {
    // TODO 3 (feature 3 — busca) + TODO 4 (feature 4 — categoria): filtrar
    // `_all` por `_categoryNames` (quando não-nulo, manter só quem está no
    // set) e por `_searchText` (substring case-insensitive do `name`) — igual
    // à lógica already implementada no gabarito (`_categoryNames!.contains`,
    // `.toLowerCase().contains`).
    return _all;
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(title: const Text('Pokédex')),
      body: Semantics(
        identifier: 'item-list-screen',
        child: _buildBody(),
      ),
    );
  }

  Widget _buildBody() {
    if (_loading) {
      return const Center(child: CircularProgressIndicator());
    }
    if (_error != null) {
      return Center(child: Text(_error!));
    }
    return Column(
      children: [
        Padding(
          padding: const EdgeInsets.all(12),
          child: Semantics(
            identifier: 'search-input',
            child: TextField(
              decoration: const InputDecoration(
                labelText: 'Buscar pokémon',
                border: OutlineInputBorder(),
                prefixIcon: Icon(Icons.search),
              ),
              onChanged: (value) => setState(() => _searchText = value),
            ),
          ),
        ),
        SizedBox(
          height: 44,
          child: ListView(
            scrollDirection: Axis.horizontal,
            padding: const EdgeInsets.symmetric(horizontal: 12),
            children: [
              _buildCategoryChip(null, 'Todos'),
              for (final category in _categories)
                _buildCategoryChip(category, category),
            ],
          ),
        ),
        const SizedBox(height: 8),
        Expanded(
          child: ListView.builder(
            itemCount: _filtered.length,
            itemBuilder: (context, index) {
              final pokemon = _filtered[index];
              return Semantics(
                identifier: 'item-card-${pokemon.id}',
                child: ListTile(
                  leading: Image.network(pokemon.spriteUrl, width: 40, height: 40),
                  title: Text(_capitalize(pokemon.name)),
                  onTap: () {
                    Navigator.of(context).push(
                      MaterialPageRoute(
                        builder: (_) => DetailScreen(pokemonId: pokemon.id),
                      ),
                    );
                  },
                ),
              );
            },
          ),
        ),
      ],
    );
  }

  Widget _buildCategoryChip(String? category, String label) {
    final selected = _selectedCategory == category;
    return Padding(
      padding: const EdgeInsets.symmetric(horizontal: 4),
      child: Semantics(
        identifier: category == null ? 'category-chip-all' : 'category-chip-$category',
        child: ChoiceChip(
          label: Text(_capitalize(label)),
          selected: selected,
          onSelected: (_) => _selectCategory(category),
        ),
      ),
    );
  }

  String _capitalize(String s) => s.isEmpty ? s : s[0].toUpperCase() + s.substring(1);
}
